pensieve/pensieve.py

```diff
@@ -172,99 +252,187 @@ def __init__(self, text, par_id=None, doc=None):
+        times.append('Par. '+str(self.id))
+        return times
+
+    def extract_activities(self, n_verbs = 5):
+        """
+        Extract verbs in the paragraph
+        """
+        verb_ranking = [(x, x.rank) for x in textacy.spacy_utils.get_main_verbs_of_sent(self.spacy_doc)]
+        verbs = []
+        for verb in sorted(verb_ranking, key=lambda x: -x[1])[:n_verbs]:
+            text = verb[0].lemma_
+            if text[0] in string.ascii_lowercase:
+                verbs.append(text)
+        return verbs
+
+    def extract_people(self):
+        """
+        Extract names in paragraph
+        """
+        names = []
+        for name in textacy.extract.named_entities(self.spacy_doc,
+                                                   include_types=['PERSON']):
+            # Exclude words too short to be names and strange characters
+            if len(name.text) < 3:
+                continue
+            # Handle possessives
+            if name.text[-2] not in string.ascii_lowercase:
+                names.append(name.text[:-2])
+            else:
+                names.append(name.text)
+        return names
+
+    def extract_places(self):
+        """
+        Extract places in paragraph
+        """
+        places = []
+        include_types = ['LOC', 'GPE', 'FACILITY']
+        for place in textacy.extract.named_entities(self.spacy_doc,
+                                                include_types=include_types):
+            if place.text != 'Hagrid':
+                places.append(place.text)
+        return places
+
+    def extract_things(self):
+        """
+        Extract things mentioned in paragraph
+        """
+        # Get named objects
+        include_types = ['ORG', 'NORP', 'WORK_OF_ART', 'PRODUCT']
+        things = []
+        for obj in textacy.extract.named_entities(self.spacy_doc,
+                                                  include_types=include_types):
+            things.append(obj)
+        # Get noun chunks
+        for nch in textacy.extract.noun_chunks(self.spacy_doc, drop_determiners=True):
+            if len(nch) == 1 and nch[0].pos == spacy.parts_of_speech.PRON:
+                continue
+            things.append(nch)
+        return things
 
-    def sanitize_places_and_objects(self, freq_cut=100):
+    def extract_mood_words(self):
+        """
+        Extract the mood/emotion of the paragraph using EMO-Lexicon
         """
-        Get a sanitized version of paragraph places and objects.
-        Get places that don't appear in the most common doc names, and
-        objects that don't appear in the most common doc places.
+        pass
 
-        NOTE: I'm not really sure what this function is actually doing...
+    def extract_mood_weights(self):
+        """
+        Extract normalized paragraph mood weights from h5 file
+        """
+        para_emotions = self.mood_weights.iloc[self.id]
+        norm = numpy.sum( para_emotions )
```

Comment:
We're suppose to absolutely normalize these vectors right?  So we don't want to normalize against itself, but norm should be the sum of the largest emotion vector I think.  It probably makes sense for norm to be a self variable in this case, where we write a function that does something like

np.argsort(np.sum(list_of_emo_vectors, axis=1))[-1] # this will give the largest sum(emo_vector)

and use this as the normalization